Preliminary support for 32-bit ARM systems

tnovak opened the pull request:
Adds basic support for ARMv7. 

The main part of this PR is the implementation of the arch interface in `arch/arm.cpp`. It also includes fixes for a few issues I've encountered while using bpftrace on my Android device running an old (4.14) armv7 kernel:
- failing to create stack maps
- errors when including kernel headers
- broken BEGIN/END probes

While many programs work on armv7 out of the box, there's one big remaining class of issues related to all pointers internally being represented as `int64`s; this can lead to unaligned loads, wrong offset calculations, etc. I'll submit a separate PR with changes to codegen to keep both easier to review.

This (aims to) resolve #1688.

##### Checklist

- [ ] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

ajor reviewed (commented):
Thanks for working on this - it's great to get support for another platform!

Could you just make sure you've covered the items in the PR-checklist please:
- Add a short line to the changelog saying you've added support for 32-bit ARM
- From the new arm.cpp I'm assuming that watchpoints work on 32-bit ARM. Can you add `ARCH armv7l` to the tests in `tests/runtime/watchpoint` if this is the case.

ajor reviewed (changes requested):
Please correct me if I'm wrong, but I think we're under-allocating in the stack maps (the problem would be truncated stack traces rather than anything awful).

From a very quick glance over other uses of uintptr_t the rest mostly look fine. I'm not 100% sure about the ones in irbuilderbpf.cpp or codegen_llvm.cpp though.

tnovak commented on src/map.cpp:
```diff
@@ -130,6 +130,14 @@ Map::Map(const SizedType &type) : IMap(type)
   int max_entries = 128 << 10;
   int flags = 0;
 
+  if ((value_size % 8) != 0)
+  {
+    // For maps of type BPF_MAP_TYPE_STACK_TRACE, value_size is expected to be
+    // a multiple of 8 (see stack_map_alloc in kernel/bpf/stackmap.c). Adjust
+    // if needed.
+    value_size = (value_size / 8 + 1) * 8;
+  }
+
```
Good catch! I assumed stack trace entries would be 4 bytes on arm, but on closer inspection `struct perf_callchain_entry` contains an `u64` array (which ultimately gets `memcpy`'d into the map): [perf_event.h](https://github.com/torvalds/linux/blob/dc164f4fb00a0abebdfff132f8bc7291a28f5401/include/linux/perf_event.h#L67).

I'll update this to `uint64_t` instead of `uintptr_t`.

tnovak commented:
Fixed the value_size calculation, updated the intptrcast test, and added a line to CHANGELOG.

Re: watchpoint tests, I haven't enabled those yet since most of them require `BPF_FUNC_send_signal` which isn't supported on my test device (running 4.14). I'm backporting that helper and will submit a follow-up for watchpoints.

From what I was able to test so far: setting a watchpoint on a userspace address works, but I'm getting `EPERM` for kernel addresses (need to debug this further).

ajor reviewed (approved):
Looks good to me!

tnovak commented on src/triggers.h:
```diff
@@ -1,7 +1,22 @@
+// Force the compiler to generate A32 instructions for the two trigger functions
+// to ensure they end up being 4-byte (UPROBE_SWBP_INSN_SIZE) aligned, otherwise
+// uprobe_register() fails with EINVAL (see kernel/events/uprobes.c).
+#define __target_attr __attribute__((target("arm")))
```
Thanks for reporting this, @daniloegea! It does look like a gcc bug; gcc 14 on [godbolt](https://gcc.godbolt.org/z/83TKac9h3) shows this:
```
<source>: In function 'int main()':
<source>:4:18: internal compiler error: canonical types differ for identical types 'void()' and 'void()'
    4 | int main() { call(f); }
      |              ~~~~^~~
```

Unfortunately, `aligned(4)` isn't sufficient -- we have to instruct the compiler to generate ARM code for these two trigger functions for uprobes to work:

- Kernel installs the probe by inserting a trap instruction ([`UPROBE_SWBP_ARM_INSN`](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/arm/include/asm/uprobes.h?h=v5.10.182#n17)). This is `udf 25` in `arch/arm`, which has a different encoding in ARM vs Thumb (`0xe7f001f9` that the kernel uses is the A32 version). With `aligned(4)`, trigger functions are still Thumb code, so calling them won't cause the CPU to switch to ARM state.
- Similarly, even if we were to force the state change (e.g. by clearing the low-order bit of the function pointer), we'd have a state mismatch after returning from the uprobe (since `bx lr` in `{BEGIN,END}_trigger` is encoded as Thumb -- `0x4770` vs `0xe12fff1e`).

I'll play around with it some more to see if there's a workaround for gcc.

tnovak replied:
@daniloegea: that might be the best solution, yeah.

This gcc issue is quite puzzling. Adding an unrelated function definition after the `target("arm")` one in your example causes it to compile fine: https://gcc.godbolt.org/z/hcoxEMs4e
